atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,304 @@
+                    return
+
+                cmd = self.get_cmd_to_write_empty_slot(slot_id)
+                _, _, exitcode = self._ssh_run_remote_cmd(client, cmd)
+
+                if exitcode != 0:
+                    logger.warning("Unlocking slot %s on host %s failed!", slot_id, self.hostname)
+            else:
+                logger.warning("%s - slot %s is available. Unlocking skipped!",
+                               self.hostname, slot_id)
+
+
+class LockedResource:
+
+    def __init__(self, host: RemoteHost, slot: int, prid: str):
+        """
+        :param slot: int, Remote host slot ID
+        :param prid: str, Pipeline run ID
+        """
+        self.slot = slot
+        self.prid = prid
+        self.host = host
+
+    def unlock(self):
+        self.host.unlock(self.slot, self.prid)
+
+
+class RemoteHostsPool:
+
+    def __init__(self, hosts: List[RemoteHost]):
+        """
+        :param hosts: List[RemoteHost], List of Remote hosts
+        """
         self.hosts = hosts
 
-    def find(self) -> List[Host]:
-        try:
-            resources = []
-            for host in self.hosts:
-                host.check_avail()
-                if len(host.free_slots) > 0:
-                    resources.append(host)
-            sorted_result = sorted(resources, key=lambda h: len(h.free_slots))
-            logger.debug('sorted result: %s', sorted_result)
-            return sorted_result
-        except Exception as err:
-            logger.debug(err)
-            logger.error("Unable to read hosts file")
-            return []
-
-    def obtain_free_resource(self, prid: str) -> Tuple[Host, int]:
-        availiable_hosts = self.find()
-        for host in availiable_hosts:
-            for sid in host.free_slots:
-                if host.lock(sid, prid):
-                    return host, sid
-        raise Exception('No availiable host found')
-
-
-if __name__ == '__main__':
-    my_hosts = [Host('username', 'hostname')]
-    mngr = ResourceManager(my_hosts)
-    # locking
-    free_host, slot = mngr.obtain_free_resource('unique-name-for-pipeline-run')
-
-    # Unlocking
-    free_host.unlock(slot, 'unique-name-for-pipeline-run')
+    @classmethod
+    def from_config(cls, config: dict):
+        """ Instantiate remote hosts loaded from configmap
+
+        :param config: dict, Arch specific remote hosts dictionary from configmap
+
+        Example:
+
+        hostname-remote-host1:
+          enabled: true
+          auth: qa-vm-secret-filepath
+          username: cloud-user
+          ...
+        hostname-remote-host2:
+          ...
+        """
+
+        # Add enabled clusters to list hosts
+        hosts = [RemoteHost(hostname=hostname, username=attr['username'], ssh_keyfile=attr['auth'])
+                 for hostname, attr in config.items() if attr['enabled']]
+
+        return cls(hosts)
+
+    @backoff.on_exception(
+        backoff.constant,
+        LockRetry,
+        max_time=TIMEOUT_FAIL_BUILD,
```

Comment:
(optional, I'm not sure here)

Reply:
@rcerven @chmeliik thoughts?

Reply:
if locking fails with retries on 1 host we should move on to another candidate host

Reply:
That already happens inside this method. If *all* slots on *all* hosts fail to lock, that's when the method raises LockRetry and does the entire process again.

I don't mind having the extra timeout here.

Reply:
ic, well there are now 2 use cases:
1) something is wrong with hosts, that we are unable to lock any resource
2) there is more pipeline runs running than remote hosts capacity

for 2nd case we should wait until there is free slot

for 1st case mmm, there might be some temporarily network issue, or there might be some serious issue with hosts, so if we will wait indefinitely (pipeline run timeout will kill run, but it will take a lot time to discover issue)